funsor/interpreter.py

```diff
@@ -324,13 +333,50 @@ def dispatched_interpretation(fn):
     return fn
 
 
+class StatefulInterpretation:
+    """
+    Base class for interpreters with instance-dependent state or parameters.
+
+    Example usage::
+
+        class MyInterpretation(StatefulInterpretation):
+
+            def __init__(self, my_param):
+                self.my_param = my_param
+
+        @MyInterpretation.register(...)
+        def my_impl(interpreter_state, cls, *args):
+            my_param = interpreter_state.my_param
+            ...
+
+        with interpretation(MyInterpretation(my_param=0.1)):
+            ...
+    """
+    def __init_subclass__(cls):
+        cls.registry = KeyedRegistry(default=lambda *args: None)
+        cls.dispatch = cls.registry.dispatch
+
+    def __call__(self, cls, *args):
+        return self.dispatch(cls, *args)(self, *args)
```

Comment:
This default implementation of `__call__()` relies on `InterpreterStack` to fall back to the previous interpretation, which is by assumption complete.